src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -2729,15 +2730,19 @@ private boolean trySetObjectField(String name, Object obj) {
     }
 
     /**
-     * Called by the VM, during -Xshare:dump
+     * Called only by the VM, during -Xshare:dump.
+     *
+     * @implNote This is done while the JVM is running in single-threaded mode,
+     * and at the very end of Java bytecode execution. We know that no more classes
+     * will be loaded and none of the fields modified by this method will be used again.
```

Comment:
I wonder if moving all this to VM side, which is not bound to Java language rules, would be conceptually cleaner. It would be more work, but I think there any only three classes that implement `resetArchivedStates`, which limits the scope of the change. I am prototyping this locally...

Reply:
Something like this: https://github.com/openjdk/jdk/compare/master...shipilev:jdk:JDK-8342283-cds-reset-archived-states. This passes `runtime/cds` at least.